python/jiminy_py/src/jiminy_py/simulator.py

```diff
@@ -519,59 +590,67 @@ def render(self,
 
         # Instantiate the robot and viewer client if necessary.
         # A new dedicated scene and window will be created.
-        if not self.is_viewer_available:
-            # Make sure that the current viewer is properly closed if any
-            if self.viewer is not None:
-                self.viewer.close()
-                self.viewer = None
-
-            # Create new viewer instance
-            self.viewer = Viewer(
-                self.robot,
-                use_theoretical_model=False,
-                open_gui_if_parent=False,
-                **viewer_kwargs)
-            assert self.viewer is not None and self.viewer.backend is not None
-
-            # Share the external force buffer of the viewer with the engine
-            if self.is_simulation_running:
-                self.viewer.f_external = [*self.robot_state.f_external][1:]
-
-            if self.viewer.backend.startswith('panda3d'):
-                # Enable display of COM, DCM and contact markers by default if
-                # the robot has freeflyer.
-                if self.robot.has_freeflyer:
-                    if "display_com" not in viewer_kwargs:
-                        self.viewer.display_center_of_mass(True)
-                    if "display_dcm" not in viewer_kwargs:
-                        self.viewer.display_capture_point(True)
-                    if "display_contacts" not in viewer_kwargs:
-                        self.viewer.display_contact_forces(True)
-
-                # Enable display of external forces by default only for
-                # the joints having an external force registered to it.
-                if "display_f_external" not in viewer_kwargs:
-                    profile_forces, *_ = self.engine.profile_forces.values()
-                    force_frames = set(
-                        self.robot.pinocchio_model.frames[f.frame_index].parent
-                        for f in profile_forces)
-                    impulse_forces, *_ = self.engine.impulse_forces.values()
-                    force_frames |= set(
-                        self.robot.pinocchio_model.frames[f.frame_index].parent
-                        for f in impulse_forces)
-                    visibility = self.viewer._display_f_external
-                    assert isinstance(visibility, list)
-                    for i in force_frames:
-                        visibility[i - 1] = True
-                    self.viewer.display_external_forces(visibility)
-
-            # Initialize camera pose
-            if self.viewer.is_backend_parent and camera_pose is None:
-                camera_pose = viewer_kwargs.get("camera_pose", (
-                    (9.0, 0.0, 2e-5), (np.pi/2, 0.0, np.pi/2), None))
+        for i, viewer in enumerate(self._viewers):
+            if not viewer.is_open():
+                viewer.close()
+                self._viewers.pop(i)
+
+        for i, robot in enumerate(self.robots):
```

Comment:
So many indents x) Maybe at this point  you should add yet another private helper method on top of this module `_create_viewer`.